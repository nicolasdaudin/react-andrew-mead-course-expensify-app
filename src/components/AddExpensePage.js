import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connectÂ } from 'react-redux';
import {addExpense} from '../actions/expenses';


export  class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
      this.props.addExpense(expense);
      this.props.history.push('/');
  }

  render() {
    return (
      <div>
        <h1>Add Expense</h1>
        <ExpenseForm onSubmit={this.onSubmit} />
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
    addExpense: (expense) => dispatch(addExpense(expense))
});


export default connect(undefined,mapDispatchToProps)(AddExpensePage);

// before adding test+jest and passing to class-based component, that was :
// the idea is to abstract away the call to addExpense, otherwise it's more complicated for jest tests
// see video 124 (section 12, vidoe 14)
// ===========
/*
const AddExpensePage = (props) => (
  <div>
    <h1>Add Expense</h1>
    <ExpenseForm onSubmit={(expense) => {

      // redirect to dashboard without a full page refresh but with a browser routing
      props.dispatch(addExpense(expense)) ;
      props.history.push('/');
    }}/>
  </div>
)
export default connect ()(AddExpensePage);*/