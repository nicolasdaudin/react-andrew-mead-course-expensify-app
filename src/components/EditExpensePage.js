import React from 'react';
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense,startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
  
  onSubmit = (expense) => {
    //console.log('updated ',expense);
    this.props.startEditExpense(this.props.expense.id,expense);
    this.props.history.push('/');
  }

  onClickRemove = () => { 
    //console.log('removed',this.props.expense.id);
    this.props.startRemoveExpense(this.props.expense.id);
    this.props.history.push('/');
  }

  render () {
    return (
      <div>
        <ExpenseForm 
          expense={this.props.expense}
          onSubmit={this.onSubmit}
        />
        <button onClick={this.onClickRemove}>Remove Expense</button>
      </div>
    )
  } 
  
}

const mapStateToProps = (state,props) => {
  // props.match is passed to us by React via component. 
  // when we did <PrivateRoute path='/edit/:id' component={EditExpensePage} /> in AppRouter, when passing EditExpensePage as 'component'
  // react inject props element with match, history and location

  return { 
    expense : state.expenses.find( (expense) => (expense.id === props.match.params.id))
  }
}

const mapDispatchToProps = (dispatch) => ({
  startEditExpense : (id,expense) => { dispatch(startEditExpense(id,expense))},
  startRemoveExpense : (id) => { dispatch(startRemoveExpense({id}))}  
});

export default connect(mapStateToProps,mapDispatchToProps)(EditExpensePage);

/* 
// see video 12.15 refactor for further testing
// that was the state before
const EditExpensePage = (props) => {
  console.log(props);
  return (
    <div>
      <ExpenseForm 
        expense={props.expense}
        onSubmit={(expense) => {
          console.log('updated ',expense);
          props.dispatch(editExpense(props.expense.id,expense));
          props.history.push('/');
        }}
      />
      <button onClick={() => { 
        props.dispatch(removeExpense({ id : props.expense.idÂ }));
        props.history.push('/');
      }}>Remove Expense</button>
    </div>
  )
}

const mapStateToProps = (state,props) => {
  return { 
    expense : state.expenses.find( (expense) => (expense.id === props.match.params.id))
  }
}

export default connect(mapStateToProps)(EditExpensePage);*/