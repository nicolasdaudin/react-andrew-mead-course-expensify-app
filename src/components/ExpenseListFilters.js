import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter,sortByAmount,sortByDate,setStartDate, setEndDate } from '../actions/filters';
import { DateRangePicker } from 'react-dates';

export class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused : null
  }

  onDatesChange = ({startDate,endDate}) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  }

  onFocusChange = (calendarFocused) => {
    this.setState(() => ({calendarFocused}))
  }

  onTextChange = (e) => { 
    this.props.setTextFilter(e.target.value)
  }
  
  onSortChange = (e) => {
    if (e.target.value === 'date') { 
      this.props.sortByDate()
    } else if (e.target.value === 'amount') {
      this.props.sortByAmount()
    }
  }

  render() {
    return (
      <div className="content-container">
        <div className="input-group">
          <div className="input-group__item">
            <input 
              type="text" 
              className="text-input" 
              value={this.props.filters.text} 
              placeholder="Search expenses"
              onChange={this.onTextChange}/>
          </div>
          <div className="input-group__item">
            <select className="select" value={this.props.filters.sortBy} onChange={this.onSortChange}>
              <option value="date">Date</option>
              <option value="amount">Amount</option>
            </select>
          </div>
          <div className="input-group__item">
            <DateRangePicker
              startDate={this.props.filters.startDate}
              endDate={this.props.filters.endDate}
              onDatesChange={this.onDatesChange}
              focusedInput={this.state.calendarFocused}
              onFocusChange={this.onFocusChange}
              showClearDates={true}
              numberOfMonths={1}
              isOutsideRange={() => false}
    
            />
          </div>
        </div>
        
       
       
          
      </div>
    )
  }
}

const mapStateToProps = (state) => ({ 
  filters : state.filters
})

const mapDispatchToProps = (dispatch) => ({
  setStartDate : (startDate) => dispatch(setStartDate(startDate)),
  setEndDate : (endDate) => dispatch(setEndDate(endDate)),
  setTextFilter : (text) => dispatch(setTextFilter(text)),
  sortByDate : () => dispatch(sortByDate()),
  sortByAmount : () => dispatch(sortByAmount())
})

const ConnectedExpenseListFilters = connect(mapStateToProps,mapDispatchToProps)(ExpenseListFilters);

export default ConnectedExpenseListFilters;