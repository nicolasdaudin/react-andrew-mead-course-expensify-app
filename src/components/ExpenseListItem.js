import React from 'react';
import moment from 'moment';
import {Link} from 'react-router-dom';
import numeral from 'numeral';

const ExpenseListItem = ({id,description,amount,createdAt}) => (
  <div className="expense">
    <strong>{description}</strong> 
    &nbsp;- Co√ªt: {numeral(amount/100).format('$0,0.00')} 
    &nbsp;- Date: {moment(createdAt).format("dddd, MMMM Do YYYY, h:mm:ss a")} 
    &nbsp;- <Link to={'/edit/' + id}>Edit {description}</Link>
  </div>
)


export default ExpenseListItem;