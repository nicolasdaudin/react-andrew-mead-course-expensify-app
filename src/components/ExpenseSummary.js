import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import selectExpensesTotal from '../selectors/expenses-total.js';
import selectExpenses from '../selectors/expenses.js';
import numeral from 'numeral'


export const ExpenseSummary = ({expenseCount,expensesTotal}) => {
  const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';
  const formattedExpensesTotal = numeral(expensesTotal / 100).format('$0,0.00');

  return (
    <div className="page-header">
      <div className="content-container">
        <h1 className="page-header__title">Viewing <span>{ expenseCount }</span> {expenseWord} for a total of <span>{formattedExpensesTotal}</span> </h1>
        <div className="page-header__actions">
          <Link className="button"to="/create">Add Expense</Link>
        </div>
      </div>
    </div>
  )
}

const mapStateToProps = (state ) => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters);
  return {
    expenseCount:visibleExpenses.length,
    expensesTotal:selectExpensesTotal(visibleExpenses)
  }
}

const ConnectedExpenseSummary = connect(mapStateToProps)(ExpenseSummary);

export default ConnectedExpenseSummary