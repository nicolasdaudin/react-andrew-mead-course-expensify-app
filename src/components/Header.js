import React from 'react';
import {NavLink} from 'react-router-dom';

const Header = () => (
  <header>
    <h1>Expensify</h1>
    <NavLink to="/" activeClassName="is-active" exact={true}>Dashboard</NavLink> -&nbsp;
    <NavLink to="/create" activeClassName="is-active">Create expense</NavLink> -&nbsp;
    <NavLink to="/help" activeClassName="is-active">Help me!</NavLink> -&nbsp;    
  </header>
)

export default Header;