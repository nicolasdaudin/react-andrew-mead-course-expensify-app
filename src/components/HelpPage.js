import React from 'react'

const HelpPage = () => (
  <div>This is my HELP expense page</div>
)

export default HelpPage;