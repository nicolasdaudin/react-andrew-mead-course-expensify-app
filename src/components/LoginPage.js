import React from 'react';
import { startLogin } from '../actions/auth';
import { connect } from 'react-redux';

export const LoginPage = ({startLogin}) => (
  <div>
    <button onClick={startLogin}>Login!!</button>
  </div>
)

const mapDispatchToProps = (dispatch) => ({
  startLogin : () => dispatch(startLogin())  
})

export default connect(undefined,mapDispatchToProps)(LoginPage);