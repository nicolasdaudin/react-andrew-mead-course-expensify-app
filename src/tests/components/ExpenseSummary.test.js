import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseSummary } from '../../components/ExpenseSummary';

test('should render ExpenseSummary correctly if no expenses',() => {
  const wrapper = shallow(<ExpenseSummary expenseCount={0} expensesTotal={0} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render ExpenseSummary correctly if one expenses or more',() => {
  const wrapper = shallow(<ExpenseSummary expenseCount={2} expensesTotal={35.21} />);
  expect(wrapper).toMatchSnapshot();
});
