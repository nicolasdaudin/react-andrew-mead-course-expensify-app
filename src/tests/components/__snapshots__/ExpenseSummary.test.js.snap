// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseSummary correctly if no expenses 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        0
      </span>
       
      expenses
       for a total of 
      <span>
        $0.00
      </span>
       
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render ExpenseSummary correctly if one expenses or more 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        2
      </span>
       
      expenses
       for a total of 
      <span>
        $0.35
      </span>
       
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;
